### ISSUE TITLE
409 Conflict while adding a new rule

### ISSUE THREAD

**Issue by aounhaider1** [eclipse/smarthome#705]
The server responds with a 409 conflict error when we try to add a new rule with a trigger of type 'TimerTrigger'. The error goes away if the type of trigger is changed to something like 'trigger' etc.

**Comment by kaikreuzer** [comment]
@aounhaider1 Could you add an example JSON that you are sending as HTTP POST?

**Comment by aounhaider1** [comment]
```json
{  
   "tags":[  

   ],
   "conditions":[  

   ],
   "description":"test description",
   "name":"Test",
   "triggers":[  
      {  
         "id":"Trigger_1",
         "label":"Timer Trigger",
         "description":"This triggers a rule based on a cron expression",
         "type":"TimerTrigger"
      }
   ],
   "actions":[  

   ]
}
```

**Comment by danchom** [comment]
The configuration property "cronExpression" of TimerTrigger is missing. It is marked as "required" into TimerTrigger module type definition and it must be available into trigger definition.

**Comment by kaikreuzer** [comment]
@danchom: But why does that result in a 409 CONFLICT? This does not sound like a correct response.

**Comment by danchom** [comment]
I can't reproduce the problem. I've tested with the above json code and add two rules with response code 201. I've receive 409 when I tried to add a rule with existing uid - but it is ok to have 409 in this case.

**Comment by aounhaider1** [comment]
That is strange! I am able to produce the 409 error with the same JSON as above and I am not passing any rule uid at all.

**Comment by danchom** [comment]
How do you create REST request?
1) I've started ESH with targetplatform/SmartHome Runtime.launch
2) I've used Advance REST Client of Chrome to send to http://localhost:8080/rest/rules post request with above JSON content. The result is always 201.

**Comment by danchom** [comment]
I created PR #756 which logs and return error message of 409 response. Please check the message

**Comment by danchom** [comment]
Is this problem appears again, after #756 fix, or it can be closed?

**Comment by kaikreuzer** [comment]
@aounhaider1 Can you answer? If the problem is solved, please close the issue.

**Comment by aounhaider1** [comment]
Yes, I can see error in log after the fix. But it still throws a 409 conflict if required configurations are missing in the request (cron expression for example). I have no issue with a 409.

**Comment by kaikreuzer** [comment]
Which means that the bug still exists and the issue should stay open.
@aounhaider1 can you confirm that you are doing exactly what @danchom describes in https://github.com/eclipse/smarthome/issues/705#issuecomment-165699162?

**Comment by aounhaider1** [comment]
1) I am starting it with targetplatform/openHAB_with_ESH.launch
2) I've tried sending request from Swagger API, Chrome rest client and Paper UI.
I get this response (which obviously goes away if I add the parameter):

```
{
  "error": {
    "message": "Creation of the rule is refused: Missing required parameter: cronExpression of type TimerTrigger",
    "http-code": 409
  }
}
```

**Comment by kaikreuzer** [comment]
Looking at the code, the RuleResource can only return 200 or 409, but nothing else:
https://github.com/eclipse/smarthome/blob/master/bundles/automation/org.eclipse.smarthome.automation.rest/src/main/java/org/eclipse/smarthome/automation/rest/internal/RuleResource.java#L100

I think a new return type should be introduced for invalid rules that are passed.

**Comment by danchom** [comment]
Now when receive error code you have a message describing the problem. The cronExpression is required parameter and I think it is definitely have to receive an error code. 
What is your proposal for error code which has to be returned when you didn't passed  required parameter for Module?

**Comment by kaikreuzer** [comment]
How about 400? https://en.wikipedia.org/wiki/List_of_HTTP_status_codes#4xx_Client_Error

**Comment by danchom** [comment]
Yes sounds good.
400 - when create a rule with not correct parameters, or missing required parameters
409 - when create a rule with UID which already exists in the RuleRegistry
Is this is ok or in the second case code 400 has to be returned again?

**Comment by kaikreuzer** [comment]
> Is this is ok

Ok!

**Comment by danchom** [comment]
Fixed In PR #830
